feat: Switch classic Kafka topic to diskless (#1673)
aiven/aiven-docs: docs/products/kafka/howto/switch-topic-to-diskless.md

- Unclean leader election must be turned off for the topic. If unclean leader
  election is enabled, the switch does not start.
- If tiered storage is not enabled for the topic, Aiven enables it automatically
  as part of the switch. Topics that cannot have tiered storage enabled, such as
  topics with compaction configured, are not eligible for the diskless switch.
- The switch runs in the background. A successful update request doesn't mean
  that every partition has finished switching. You can't view per-partition
  switch status or progress in the Aiven CLI, Aiven API, topic configuration,
  or customer-facing metrics integrations.

## Switch a topic to a diskless topic

To switch a classic topic to a diskless topic, enable diskless for the topic using the
Aiven CLI or Aiven API.

<Tabs groupId="switch-method">
<TabItem value="cli" label="Aiven CLI" default>

Run the following command:

```bash
avn service topic-update SERVICE_NAME TOPIC_NAME \
  --project PROJECT_NAME \
  --diskless-enable
```

Replace the following values:

- `PROJECT_NAME`: Name of your Aiven project.
- `SERVICE_NAME`: Name of your Aiven for Apache Kafka service.
- `TOPIC_NAME`: Name of the topic to switch.

> - `PARTITION_COUNT`: Current number of partitions in the topic. Use the existing

Reviewer: ```suggestion
```

>   partition count. Do not change the partition count as part of the switch.

Reviewer: ```suggestion
```

---
title: Switch a classic topic to a diskless topic
sidebar_label: Switch to a diskless topic
early: true
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
import RelatedPages from "@site/src/components/RelatedPages";

Switch an existing
[classic topic](/docs/products/kafka/diskless/concepts/topics-vs-classic)
in Aiven for Apache Kafka® to a diskless topic without copying data or renaming
the topic. The topic remains available during the switch.

Records written before the switch remain readable from the classic topic log.
Aiven writes new records to the diskless topic.

:::note
This feature is in
[early availability](/docs/platform/concepts/service-and-feature-releases#early-availability-)
and is not enabled by default. To request access, contact your account team or
[Aiven support](/docs/platform/howto/support#create-a-support-ticket).
:::

## Prerequisites

> - The classic-to-diskless topic switch feature and

Reviewer 2: @Stuzanna wonder if here we should call it: "Tiered Storage unification feature" instead.. preparing the ground for upcoming changes, wdyt?

Also, how would the user know if this feature is enabled? Maybe it's just ok to trust us when requested, but I don't think there's anything observable for users apart from metrics.

  [diskless topics](/docs/products/kafka/diskless/concepts/diskless-topic-overview)
  are enabled for the service.
- The service runs Apache Kafka® 4.1 or later.
- [Tiered storage](/docs/products/kafka/howto/configure-topic-tiered-storage)

>   is enabled for the topic before you switch it to diskless.

Reviewer 2: I think we are mixing 2 things here: pre-req for enabling the feature on the service and the pre-req to apply the change to a topic. 
This is also not a pre-req, but something that happens as a topic is switched: a classic topic without TS enabled is switched to diskless: then diskless and remote-storage are enabled on the same transaction.
We can mention that if classic topics don't have remote-storage enabled, then it will be enabled as part of the switch, and that if they are not able to switch remote-storage=on (e.g. compact topics) then they are not eligible for diskless.

But maybe better to move the topic-specifics to "considerations" section below?

Author: Thanks, updated to separate service prerequisites from topic considerations and clarify the switch behavior.

  partition count. Do not change the partition count as part of the switch.

</TabItem>
<TabItem value="api" label="Aiven API">

Send a request to update the topic configuration:

```bash
API_URL="https://api.aiven.io/v1/project/PROJECT_NAME/service"
API_URL="${API_URL}/SERVICE_NAME/topic/TOPIC_NAME"

curl --request PUT \
  --url "${API_URL}" \
  --header "Authorization: Bearer TOKEN" \
  --header "Content-Type: application/json" \
  --data '{
    "config": {
      "diskless_enable": true
    }
  }'
```

Replace the following values:

- `PROJECT_NAME`: Name of your Aiven project.
- `SERVICE_NAME`: Name of your Aiven for Apache Kafka service.
- `TOPIC_NAME`: Name of the topic to switch.
- `TOKEN`: Your Aiven API token.

</TabItem>
</Tabs>

After the update request succeeds, Aiven starts switching the topic to a
diskless topic.

## Verify the switch request

Verify the topic configuration to confirm that Aiven accepted the diskless
switch request.

<Tabs groupId="confirm-method">
<TabItem value="cli" label="Aiven CLI" default>

The command uses `jq` to filter the output for the selected topic.

Run the following command:

```bash

> avn service topic-list SERVICE_NAME \

Reviewer 2: would `avn service topic-get...` work better here? IIRC, topic list includes configs but topic-get does not, which doesn't make much sense to me. Maybe this should be fixed? (not-blocker)

Author: Good point. I tested this again on dev-sandbox. `topic-get` only includes partition and consumer group details, not the config fields needed for this verification.

`topic-list --json | jq` returns `diskless_enable=true` and `remote_storage_enable=true`, so I used it to confirm that Aiven accepted the switch request. Let me know if you’d still prefer `topic-get` here.

Reviewer 2: Yeah, that's not good. I'm extending the topic-get command: https://github.com/aiven/aiven-client/pull/467 -- let's see if we can land it soon enough to include it.

Reviewer 2: I've released 4.16, we can use topic-get if users update their CLI: https://github.com/aiven/aiven-client/releases/tag/4.16.0

Author: Thanks! Updated the command :)

  default producer retry settings, no special tuning is usually required.
- Consumer applications do not need changes to read records written before or
  after the switch.
- The topic name and retention settings do not change.
- Aiven manages the partition-level switch process. You do not need to take
  action after the update request succeeds.

### Producer settings

If you changed any of these Kafka producer settings, verify that the values
allow producers to retry records for at least as long as the defaults:

| Producer setting | Default | Why it matters during the switch |
| --- | --- | --- |
| `delivery.timeout.ms` | `120000` | Allows up to 2 minutes for retrying a record. |
| `retries` | `2147483647` | Effectively unlimited, but bounded by `delivery.timeout.ms`. |
| `enable.idempotence` | `true` | Avoids duplicate or reordered records. |
| `acks` | `all` | Supports `enable.idempotence` and durable failover. |

## How the switch works

When you switch a topic to a diskless topic, Aiven does the following for each
partition:

1. Stops writing new records to the classic topic log.
1. Waits until records written to the classic topic log are safely

>    replicated.

Reviewer 2: ```suggestion
1. Waits until records written to the classic topic partitions are safely
   replicated.
```

> 1. Records the offset where the diskless topic starts.

Reviewer 2: ```suggestion
1. Records the offsets where the diskless topic partitions start.
```

> 1. Initializes the diskless topic from that offset.

Reviewer 2: ```suggestion
1. Initializes the diskless topic partitions from those offsets.
```

> 1. Routes new writes to the diskless topic.

Reviewer 2: ```suggestion
1. Reads and writes from those topics are now handle by the Diskless storage system.
```

import TabItem from '@theme/TabItem';
import RelatedPages from "@site/src/components/RelatedPages";

Switch an existing [classic topic](/docs/products/kafka/diskless/concepts/topics-vs-classic) in Aiven for Apache Kafka® to a diskless topic.
You do not need to copy data or rename the topic. The topic remains available during the
switch.

Records written before the switch remain readable from the classic topic log,
and new records are written to the diskless topic.

Switching a topic to diskless applies only to Aiven for Apache Kafka services
that have
[diskless topics](/docs/products/kafka/diskless/concepts/diskless-topic-overview)
enabled.

:::note
This feature is in
[early availability](/docs/platform/concepts/service-and-feature-releases#early-availability-)
and is not enabled by default. To request access, contact your account team or
[Aiven support](/docs/platform/howto/support#create-a-support-ticket).
:::

## Prerequisites

- [Diskless topics](/docs/products/kafka/diskless/concepts/diskless-topic-overview)
  are enabled for the service.
- The service runs Apache Kafka® 4.1 or later.
- [Tiered storage](/docs/products/kafka/howto/configure-topic-tiered-storage) is

>   enabled for the topic before you switch it to diskless.

Reviewer 3: Same here: https://github.com/aiven/aiven-docs/pull/1673/changes#r3492692325
We can remove this

>   enabled for the topic before you switch it to diskless.

Reviewer 2: isn't this a given? maybe still worth mentioning, but as TS is pre-requisite to diskless, it can be removed?

Reviewer: BYOC is it also enforced on at the service level?

Reviewer 2: yes, feature-wise we cannot enable diskless without enabling tiered storage first.

- Unclean leader election is turned off for the topic. If unclean leader election
  is enabled, the switch does not start.
- You have access to one of the following:

  - [Aiven CLI](/docs/tools/cli) installed and authenticated.
  - [Aiven API token](/docs/platform/howto/create_authentication_token) for API
    requests.

:::note

> Enable tiered storage before you switch the topic to diskless. Do not enable

Reviewer: Lets confirm this, if you do need to do 2-step, then let's have that as part of at least one example, or, link to how to enable ts on the topic.

Reviewer 3: As we discussed before, this is not necessary. We can remove this entire note, because for the managed version of Inkless (so not the open source one), the customer can just flip diskless.enable=true and tiered storage will also automatically enabled.

Reviewer 2: Agree, however there is a fix related to this that should be shipped first for this to be true: https://aiven.atlassian.net/browse/KC-293

tiered storage and diskless in the same topic update request. This requirement
also applies to Bring Your Own Cloud (BYOC) services with diskless topics.
:::

## Considerations

Before switching a topic, review the following:

- The switch is one-way. You cannot switch a diskless topic back to a classic

>   topic.

Reviewer 2: worth mentioning that support for this is planned to land soon?

Reviewer: ```suggestion
- The switch is one-way. You cannot switch a diskless topic back to a classic
  topic, at this time.
```

Reviewer: Not sure we want to say Soon, but agree implying we could look at this.

- Diskless topic limitations apply after the switch. Review
  [Limitations of diskless topics](/docs/products/kafka/diskless/concepts/limitations)
  to confirm that diskless topics support your workload.
- The switch runs in the background. Internal partition-level switch state is not

>   exposed in the Aiven CLI, Aiven API, or topic configuration.

Reviewer 2: this is not clear from user pov. can we elaborate this further?

Reviewer: So the state of the switch is not exposed

Reviewer 2: Is it exposed via metrics? Wonder if we should expose the offsets as metrics.. IIRC log start/end offsets are exposed.

Author: Updated this to clarify the user-facing impact: a successful update request doesn’t mean every partition has finished switching, and per-partition switch status/progress isn’t exposed in the CLI, API, topic configuration, or customer-facing metrics integrations.

## Switch a topic to a diskless topic

To switch a classic topic to a diskless topic, enable diskless for the topic using the
Aiven CLI or Aiven API.

:::note
The topic keeps its existing tiered storage setting. Do not include
`remote_storage_enable` in the diskless update request. If you set

> `remote_storage_enable` and `diskless_enable` in the same request, the request

Reviewer 2: this should not be the case with the feature flag enabled. if they happen to enable both configs at once it should be supported. the only unsupported case is updating with diskless enabled but remote storage disabled (forbidden state)

Author: Tested this on  dev-sandbox on a service with the feature flag enabled: setting both remote_storage_enable=true and diskless_enable=true in one request returns this error: “Tiered Storage and Diskless feature cannot be enabled together on the same topic.”

Reviewer 2: Good catch. Fix: https://github.com/aiven/aiven-core/pull/76142

Author: Removed this note.

returns an error.
:::

<Tabs groupId="switch-method">
<TabItem value="cli" label="Aiven CLI" default>

Run the following command:

```bash
avn service topic-update SERVICE_NAME TOPIC_NAME \
  --project PROJECT_NAME \

>   --partitions PARTITION_COUNT \

Reviewer 3: For now this is required, but we need to understand why, it really should not be @Stuzanna

Reviewer: Partition count you mean?

Reviewer 3: yes, it should not be required

Reviewer 2: should be fixed: https://github.com/aiven/aiven-client/pull/460 but not sure if it has been released.

Reviewer: Ah we hadn't seen, linked to internal ticket and closed.

Reviewer: ```suggestion
```

- `PROJECT_NAME`: Name of your Aiven project.
- `SERVICE_NAME`: Name of your Aiven for Apache Kafka service.
- `TOPIC_NAME`: Name of the topic to switch.
- `PARTITION_COUNT`: Current number of partitions in the topic. Use the existing
  partition count. Do not change the partition count as part of the switch.

</TabItem>
<TabItem value="api" label="Aiven API">

Send a request to update the topic configuration:

```bash
API_URL="https://api.aiven.io/v1/project/PROJECT_NAME/service"
API_URL="${API_URL}/SERVICE_NAME/topic/TOPIC_NAME"

curl --request PUT \
  --url "${API_URL}" \
  --header "Authorization: Bearer TOKEN" \
  --header "Content-Type: application/json" \
  --data '{
    "config": {
      "diskless_enable": true
    }
  }'
```

Replace the following values:

- `PROJECT_NAME`: Name of your Aiven project.
- `SERVICE_NAME`: Name of your Aiven for Apache Kafka service.
- `TOPIC_NAME`: Name of the topic to switch.
- `TOKEN`: Your Aiven API token.

</TabItem>
</Tabs>

After the update request succeeds, Aiven starts switching the topic to a
diskless topic.

## Verify the topic configuration

Verify the topic configuration to confirm that the diskless switch request was

> accepted and that tiered storage remains enabled.

Reviewer 3: Let's remove mention of tiered storage in this section, I think it's confusing

both set to `true`. This confirms that the diskless switch request was accepted.
The partition-level switch can continue in the background and is not exposed in
the topic configuration.

</TabItem>
<TabItem value="api" label="Aiven API">

Send a request to get the topic configuration:

```bash
API_URL="https://api.aiven.io/v1/project/PROJECT_NAME/service"
API_URL="${API_URL}/SERVICE_NAME/topic/TOPIC_NAME"

curl --request GET \
  --url "${API_URL}" \
  --header "Authorization: Bearer TOKEN"
```

Replace the following values:

- `PROJECT_NAME`: Name of your Aiven project.
- `SERVICE_NAME`: Name of your Aiven for Apache Kafka service.
- `TOPIC_NAME`: Name of the topic.
- `TOKEN`: Your Aiven API token.

In the response, verify that `diskless_enable` and `remote_storage_enable` are
both set to `true`. This confirms that the diskless switch request was accepted.
The partition-level switch can continue in the background and is not exposed in
the topic configuration.

</TabItem>
</Tabs>

## What to expect during the switch

During the switch:

- The topic remains available.
- Producers might briefly receive retriable errors while the topic is switched.

>   Most Kafka clients retry these errors by default.

Reviewer 2: let's probably suggest them to have these configs in place, e.g. max retries at long.max_value, and long enough delivery timeout.

Reviewer: agreed

Author: Thanks, agreed. Can you confirm the exact producer configs and values we should recommend here? I have `retries` and `delivery.timeout.ms`, but want to confirm the correct Kafka config names and whether we should give specific values or keep the guidance general.

Reviewer 2: defaults should be good. So we should recommend updating in case users are already changing these values. e.g. maybe a bit too verbose, but:
```suggestion
  Most Kafka clients retry these errors by default, and the default producer settings already tolerate this window, so no special tuning is required.
  
If you have overridden any of the following, reconsider lowering them below their defaults is what causes these transient errors to surface as failures:

| Config                | Default      | Why it matters during a switch                                  |
| --------------------- | ------------ | --------------------------------------------------------------- |
| `delivery.timeout.ms` | `120000`     | 2 min upper bound on retrying a record; far exceeds the switch. |
| `retries`             | `2147483647` | Effectively unlimited; bounded by `delivery.timeout.ms`.        |
| `enable.idempotence`  | `true`       | Keeps retries safe — no duplicates or reordering.               |
| `acks`                | `all`        | Required for idempotence; ensures durability during failover.   |
```

Author: Thanks, updated. I added the producer defaults to check if users have changed these settings, and kept the table in a separate Producer settings section for readability.

early: true
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
import RelatedPages from "@site/src/components/RelatedPages";

Switch an existing
[classic topic](/docs/products/kafka/diskless/concepts/topics-vs-classic)
in Aiven for Apache Kafka® to a diskless topic without copying data or renaming
the topic. The topic remains available during the switch.

Records written before the switch remain readable from the classic topic log.
Aiven writes new records to the diskless topic.

:::note
This feature is in
[early availability](/docs/platform/concepts/service-and-feature-releases#early-availability-)
and is not enabled by default. To request access, contact your account team or
[Aiven support](/docs/platform/howto/support#create-a-support-ticket).
:::

## Prerequisites

- The [diskless topics](/docs/products/kafka/diskless/concepts/diskless-topic-overview)
  feature is enabled for the service. To request access, contact your account team or
  [Aiven support](/docs/platform/howto/support#create-a-support-ticket).
- The service runs Apache Kafka® 4.1 or later.
- Unclean leader election is turned off for the topic. If unclean leader election

>   is enabled, the switch does not start.

Reviewer 2: I think this one belongs to topic considerations as well
